controllers/volumereplicationgroup_controller.go

```diff
@@ -753,55 +779,52 @@ func (v *VRGInstance) processAsPrimary() (ctrl.Result, error) {
 		return ctrl.Result{Requeue: true}, nil
 	}
 
-	if err := v.restorePVs(); err != nil {
+	result := ctrl.Result{}
+	if err := v.restorePVs(&result); err != nil {
 		v.log.Info("Restoring PVs failed", "Error", err.Error())
 
 		msg := fmt.Sprintf("Failed to restore PVs (%v)", err.Error())
 		setVRGClusterDataErrorCondition(&v.instance.Status.Conditions, v.instance.Generation, msg)
 
 		if err = v.updateVRGStatus(false); err != nil {
 			v.log.Error(err, "VRG Status update failed")
+
+			result.Requeue = true
 		}
 
-		// Since updating status failed, reconcile
-		return ctrl.Result{Requeue: true}, nil
+		return result, nil
 	}
 
-	requeue := v.handleVRGMode(ramendrv1alpha1.Primary)
+	result = v.reconcileAsPrimary()
 
 	// If requeue is false, then VRG was successfully processed as primary.
 	// Hence the event to be generated is Success of type normal.
 	// Expectation is that, if something failed and requeue is true, then
 	// appropriate event might have been captured at the time of failure.
-	if !requeue {
+	if !result.Requeue {
 		rmnutil.ReportIfNotPresent(v.reconciler.eventRecorder, v.instance, corev1.EventTypeNormal,
 			rmnutil.EventReasonPrimarySuccess, "Primary Success")
 	}
 
 	if err := v.updateVRGStatus(true); err != nil {
-		requeue = true
-	}
-
-	if requeue {
-		v.log.Info("Requeuing resource as Primary")
-
-		return ctrl.Result{Requeue: true}, nil
+		result.Requeue = true
 	}
 
-	v.log.Info("Successfully processed vrg as primary")
+	delayResetIfRequeueTrue(&result, v.log)
 
-	return ctrl.Result{}, nil
+	return result, nil
 }
 
-func (v *VRGInstance) reconcileAsPrimary() bool {
-	requeueForVolSync := false
+func (v *VRGInstance) reconcileAsPrimary() ctrl.Result {
+	result := ctrl.Result{}
 	if len(v.volSyncPVCs) != 0 {
-		requeueForVolSync = v.reconcileVolSyncAsPrimary()
+		result.Requeue = v.reconcileVolSyncAsPrimary()
```

Comment:
Another issue with VolSync. If VolSync needs to requeue, VolRep or ShioRamen may turn the result off.  The removed line (#804) is needed.  It should be updated to:
```
result.Requeue = requeueForVolSync || requeueForVolRep || requeueForKubeObjectProtection
return result
```

Reply:
The intent was to:
- initialize `result.Requeue` to false
- set `result.Requeue` to whatever `v.reconcileVolSyncAsPrimary` returns
- pass `result.Requeue` as a pointer to `v.reconcileVolRepsAsPrimary` which may set it to `true` only, but never `false`
- pass `result` as a pointer to `v.kubeObjectsProtect` which may set it to `true` only, but never `false`

I don't know of a way `result.Requeue` can change to `false` after being set to `true` by `v.reconcileVolSyncAsPrimary`. 
 I'd be curious to see the logs or code showing how this issue can be produced.

I suppose this could be changed back to the way it was with each subcomponent returning its `requeue` boolean and ORing them together.  `v.kubeObjectsProtect` uses the result.RequeueAfter in addition to the `result.Requeue` boolean.

Reply:
In my opinion this comes down to trust.  If the volRep and kubeObjects subcomponents are trusted to only set `Requeue` to `true`, then no change is needed.  If they are not trusted, then additional logic to set bool to logical OR of each can be added.  For now assuming subcomponents can be trusted.

Reply:
I wholeheartedly agree with you @hatfieldbrian.  It all comes down to trust. At first glance, I didn't trust that those two calls do not modify the result pointer. So I just assumed that one of them did.  but it turned out that the `requeue` was modified in my throw-away code trying to fix issue 2. This code can definitely stay the way it is. However, this trust is not strong enough. Tomorrow someone will explicitly set the requeue to false, and then we fall in the same problem.  Anyway, thanks for the update.  I'll give it a try.